src/os_posix/os_fs.c

```diff
@@ -789,3 +971,175 @@ __wt_os_posix(WT_SESSION_IMPL *session)
+    while ((old_state = pfh->mmap_resizing) == 1)
+        WT_PAUSE();
+
+    if ((cas_result = __wt_atomic_casv32(&pfh->mmap_resizing, old_state, 1)) == false)
+        goto wait;
+
+    /*
+     * Wait for any sessions using the region for I/O to finish. Now that we have set the resizing
+     * flag, new sessions will not use the region, defaulting to system calls instead.
+     */
+    while (pfh->mmap_usecount > 0)
+        WT_PAUSE();
+}
+
+/*
+ * This LWN article (https://lwn.net/Articles/731706/) describes a potential problem when mmap is
+ * used over a direct-access (DAX) file system. If a new block is created and then the file is
+ * memory-mapped and the client writes to that block via mmap directly into storage (via DAX),
+ * the file system may not know that the data was written, so it may not flush the metadata
+ * prior to data being written. Therefore, the block may be reallocated or lost upon crash.
+ *
+ * WiredTiger currently disallows using the mmap option with the direct I/O option. We are relying
+ * on the user correctly specifying the direct I/O option if they mount a file system as DAX. If
+ * we did not wish to rely on the user supplying the correct flags, we have two options:
+ *
+ * (1) Use MAP_SYNC flag available on some versions of Linux. The downside is being Linux-specific
+ *     and not extensively tested (this is a recent flag).
+ *
+ * (2) Always fsync when we unmap the file. In our implementation, if a session extends the file by
+ *     writing a new block beyond the current file size, we always unmap the file and then re-map it
+ *     before allowing any reads or writes via mmap into the new block. If we sync the file upon
+ *     unmapping, we will be certain that the metadata is persistent.
+ */
+/*
+ * __wt_map_region --
+ *     Map the virtual address region backed by a file into our address space.
+ */
+int
+__wt_map_region(WT_FILE_HANDLE *file_handle, WT_SESSION *wt_session)
+{
+    WT_FILE_HANDLE_POSIX *pfh;
+    WT_SESSION_IMPL *session;
+    wt_off_t file_size;
+    void *previous_address;
+
+    session = (WT_SESSION_IMPL *)wt_session;
+    pfh = (WT_FILE_HANDLE_POSIX *)file_handle;
+
+    WT_ASSERT(session, S2C(session)->mmap);
+    WT_ASSERT(session, pfh->mmap_file_mappable);
+    WT_ASSERT(session, pfh->mmap_prot);
+
+    if (__posix_file_size((WT_FILE_HANDLE *)pfh, wt_session, &file_size))
+        WT_RET_MSG(session, __wt_errno(), "%s: __posix_file_size", file_handle->name);
+
+    if (file_size <= 0) {
+        if (pfh->mmap_buf != NULL)
+            __wt_unmap_region(file_handle, wt_session);
+        return (0);
+    }
+
+    /* If the buffer was previously mapped, try to remap it to the same address */
+    previous_address = pfh->mmap_buf;
+    if ((pfh->mmap_buf = (char *)mmap(previous_address, (size_t)file_size, pfh->mmap_prot,
+           MAP_SHARED | MAP_FILE, pfh->fd, 0)) == MAP_FAILED) {
+        pfh->mmap_size = 0;
+        pfh->mmap_buf = NULL;
```

Comment:
Why do we care if the file is mapped at the same address?

Reply:
There should be fewer page faults and TLB faults if we remap it at the same address.

Reply:
Yes, but there's no reason to believe we can remap at the chosen address (the file grew and it won't fit). What happens if we can't remap at the same address?

Ummm, further I think this implies we have to be able to seamlessly switch back and forth between mmap & not-mmap. The mmap system call can fail for lots of reasons, and we shouldn't crash the database if that happens, should we?

Reply:
If the file cannot be remapped at the same address, it'll be mapped at a different one. This is a suggestion, not a directive to the system call. If I used MAP_FIXED flag, then indeed the system call would fail if it could not remap at the same address, but I am not using MAP_FIXED.

Reply:
I've now implemented it such that we don't fail if we cannot mmap the file. We simply mark this file as "not mappable", so future I/O will use the system call route.